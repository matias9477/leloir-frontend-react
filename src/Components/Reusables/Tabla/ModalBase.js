import React from 'react';
import {Button, Segment} from 'semantic-ui-react';

import "./modal.css";

export const Modal = ({handleClose, show, children}) => {

    const showHideClassName = show ? "modal display-block" : "modal display-none";

    return (
        <div className={showHideClassName}>
            <Segment raised compact className="modal-main">
                <Button 
                inverted
                color='red'
                floated='right' size='tiny'
                onClick={handleClose}>X</Button>
                
                <Segment>
                    {children}
                </Segment>

            </Segment>
        </div>
    );
};