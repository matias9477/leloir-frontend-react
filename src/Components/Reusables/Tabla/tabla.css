.filter {
    margin-top: 3em;
}

.filter .ui.dropdown {
    right: 35px;
    font-size: 1.1em;
}

.filter .ui.input {
    width: 200%;
}

.tabla {
    margin-left: auto;
    margin-right: auto;
    width: 80%;
    margin-bottom: 20px;
    margin-top: 5em;
}

.tabla .ui.header {
    font-family: 'Playfair Display', serif;
    font-size: 2.8em;
    text-transform: uppercase;
    font-weight: 300;
}

.tabla .ui.button {
    background-color: transparent;
}

.listadosBaja {
    color: rgb(163, 160, 160);
}

.importeNegativo {
    color: red;
}

.expansibleSeccion {
    box-shadow: 0px 0px 100px -26px rgba(0,0,0,0.73) !important;
    padding: 1.2rem !important;
    margin-bottom: 1.1rem !important;
}