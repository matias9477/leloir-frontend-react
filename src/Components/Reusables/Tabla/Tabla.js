import React from 'react';
import { Header, Pagination, Input, Dropdown, Grid, Button, Icon } from 'semantic-ui-react';
import {Link } from 'react-router-dom'

import { nroPorPagina } from '../../../Constants/utils'
import { getHumanDate } from '../../../Services/MetodosPaciente';
import './tabla.css';


class Tabla extends React.Component {
    constructor(props){
        super(props)
        this.state = {
            limit: nroPorPagina[1].value,
            activePage: 1,
            totalCount: props.data.length,
            sortParams:{
                direction: 'desc',
                param: props.param,
            }, 
            filter: '',
            param: props.param,
        }
      }

    entriesPerPage() {
        return (
            <Dropdown
            inline
            options={nroPorPagina}
            value = {this.state.limit}
            onChange={this.limitChange} 
            />          
        )
    }

    limitChange = (e, data) => {
        this.setState({
            limit: data.value,
            activePage: 1,
        })
        return this.handleSearch(((this.state.activePage - 1) * this.state.limit), (this.state.activePage * this.state.limit))
    }

    onChangePage = (e, {activePage}) => {
        if (activePage === this.state.activePage){
            return null
        } else {
            this.setState({
                activePage,
            })
            return (this.handleSearch(((this.state.activePage-1) * this.state.limit), (this.state.activePage * this.state.limit) ))
        }        
    }
    
    handleColumnHeaderClick(sortKey) { 
        this.setState({
            param: sortKey
        })

        if (this.state.sortParams.direction === 'desc'){
            this.setState({
                sortParams: {
                    direction: 'asc'
                }
            })
        } else {
            this.setState({
                sortParams: {
                    direction: 'desc'
                }
            })
        }          
    } 

    filtado(array){
        let { param } = this.state

        if (this.state.sortParams.direction === 'desc'){
            return array.sort((a, b) => (a[param] > b[param]) ? -1 : 1)
        } else {
            return array.sort((a, b) => (a[param] > b[param]) ? 1 : -1)
        }           
    }

    handleSearch = (from, to) => {
        if (this.state.filter === ''){
            return this.filtado(this.props.data).slice(from, to)
        } 
        else {

            let {filter} = this.state
            let filteredData = []

            for (let i = 0; i < this.props.data.length; i++) {
                for (let index = 0; index < this.props.columns.length; index++) {
                    var a = this.props.columns[index].dataField
                    var b = this.props.data[i]
                
                    if (b[a]===undefined || b[a]===null) {
                        break
                    } else if (b[a].toString().toUpperCase().includes(filter.toString().toUpperCase()) === true) {
                        filteredData.push(this.props.data[i])
                        break
                    }
                }
            }

            if (this.state.totalCount !== filteredData.length) {
                this.setState({
                    totalCount: filteredData.length,
                })
            }

            if(this.state.activePage > Math.ceil(this.state.totalCount / this.state.limit)){
                this.setState({
                    activePage: Math.ceil(this.state.totalCount / this.state.limit)
                })
            }

            return this.filtado(filteredData).slice(from, to)
        }
    }

    checkEmptyEntry(value){
        if(value === '' || value === undefined || value === null){
            return '-'
        } 
        return value
        
    }

    actualState(bitAlta){
        if(bitAlta){
            return "Dar de baja"
        }
        else {
            return "Dar de alta"
        }
    }

    confirmationSwitchAltaMessage(obj){
        if (obj.bitAlta){
            return (`¿Esta seguro que lo quiere dar de baja?`)
        }
        else {
            return (`¿Esta seguro que quiere lo quiere dar de alta?`)
        }
    }


    getIconTipo(tipo){
        if (tipo === 'ANIMAL'){
            return 'paw'
        } else if(tipo === 'PERSONA'){
            return 'user'
        } else if(tipo === 'INSTITUCION'){
            return 'building'
        }
    }

    render(){
        const { param } = this.props

        return <div className='tabla'>
            
            <Header as='h2'>{this.props.title}</Header>
            
            <div className='filter'>
                <Grid>
                    <Grid.Column floated='left' width={4}>
                        <Input value={this.state.filter} 
                            onChange={(filter)=>
                                this.setState({
                                    filter: filter.target.value
                                })}
                                    
                            placeholder='Ingrese búsqueda...' icon={{name: 'search'}} 
                        />
                    </Grid.Column>
                    <Grid.Column floated='right'>
                        {this.entriesPerPage()}
                    </Grid.Column>
                </Grid>
            </div>
                
            <table className='ui single line table' >
                <thead className='centerAlignment'>
                    <tr>
                        {this.props.columns.map((col, index) => ( 
                            <th key={index} onClick={() => this.handleColumnHeaderClick(col.dataField)}>{col.text}</th>
                        ))}
                        {this.props.path ? <th/> : null}
                    </tr>
                </thead>
            
                <tbody className='centerAlignment'>
                
                    {(this.handleSearch(((this.state.activePage-1) * this.state.limit), (this.state.activePage * this.state.limit))).map(  (row, index) => (

                        <tr key={index} value={row}  className={(row.bitAlta===false ) ? "listadosBaja" : null}> 

                            {this.props.columns.map((col, index)=> (
                                <td key={index} data-label={col.text}>
                                    {col.type === 'Date' ? getHumanDate(row[col.dataField]) : (col.type ==='icon' ?  <Icon name={this.getIconTipo(row[col.dataField])}/> : this.checkEmptyEntry(row[col.dataField]))}
                                </td> 
                            ))}

                            {this.props.options ? <td>
                                {this.props.path ?
                                    <Button circular icon='settings'
                                    as= {Link} 
                                    to={{pathname: `${this.props.path}${this.props.data[index][param]}`, state: { prevPath: window.location.pathname }}} 
                                    exact='true' style={{marginLeft: 'auto', marginRight: 'auto'}}
                                    >
                                    </Button> 
                                : null }
                            </td>  : null}
                    
                        </tr>
                    ))}
                    
                </tbody>
            
            </table>
            <Pagination
                totalPages={Math.ceil(this.state.totalCount / this.state.limit)}
                activePage={(this.state.activePage > Math.ceil(this.state.totalCount / this.state.limit) ? Math.ceil(this.state.totalCount / this.state.limit) : this.state.activePage)}
                onPageChange={this.onChangePage}
            />
            
        </div>
        
    }

}


export default Tabla