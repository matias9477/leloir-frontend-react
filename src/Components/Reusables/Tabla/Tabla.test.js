import React from 'react';
import { shallow } from 'enzyme';

import Tabla from './Tabla';

const pacientes = {}

it('should render the component correctly with the given mocked props', ()=>{
    const component = shallow(<Tabla
                                data={pacientes}
                                param={'id'}
                                columns={columns}
                                title='Pacientes'
                                options={true}
                                path={'/pacientes/consulta/'}
                                dataConsulta={"tipoPaciente"}/>)

    expect(component).toMatchSnapshot();
});