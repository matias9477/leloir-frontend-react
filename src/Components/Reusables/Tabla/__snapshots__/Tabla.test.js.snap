// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the component correctly with the given mocked props 1`] = `
<div
  className="tabla"
>
  <Grid
    columns={3}
  >
    <GridColumn />
    <GridColumn>
      <Header
        as="h2"
      >
        Pacientes
      </Header>
    </GridColumn>
    <GridColumn />
  </Grid>
  <div
    className="filter"
  >
    <Grid>
      <GridColumn
        floated="left"
        width={4}
      >
        <Input
          icon={
            Object {
              "name": "search",
            }
          }
          onChange={[Function]}
          placeholder="Ingrese búsqueda..."
          type="text"
          value=""
        />
      </GridColumn>
      <GridColumn
        floated="right"
      >
        <Dropdown
          additionLabel="Add "
          additionPosition="top"
          closeOnBlur={true}
          closeOnEscape={true}
          deburr={false}
          icon="dropdown"
          inline={true}
          minCharacters={1}
          noResultsMessage="No results found."
          onChange={[Function]}
          openOnFocus={true}
          options={
            Array [
              Object {
                "key": 10,
                "text": "10",
                "value": 10,
              },
              Object {
                "key": 25,
                "text": "25",
                "value": 25,
              },
              Object {
                "key": 50,
                "text": "50",
                "value": 50,
              },
              Object {
                "key": 100,
                "text": "100",
                "value": 100,
              },
            ]
          }
          renderLabel={[Function]}
          searchInput="text"
          selectOnBlur={true}
          selectOnNavigation={true}
          value={25}
          wrapSelection={true}
        />
      </GridColumn>
    </Grid>
  </div>
  <table
    className="ui single line table"
  >
    <thead
      className="centerAlignment"
    >
      <tr>
        <th
          key="0"
          onClick={[Function]}
        >
          Fecha Alta
        </th>
        <th
          key="1"
          onClick={[Function]}
        >
          Tipo
        </th>
        <th
          key="2"
          onClick={[Function]}
        >
          Paciente
        </th>
        <th
          key="3"
          onClick={[Function]}
        >
          Documento
        </th>
        <th />
      </tr>
    </thead>
    <tbody
      className="centerAlignment"
    />
  </table>
  <Pagination
    activePage={0}
    aria-label="Pagination Navigation"
    boundaryRange={1}
    ellipsisItem="..."
    firstItem={
      Object {
        "aria-label": "First item",
        "content": "«",
      }
    }
    lastItem={
      Object {
        "aria-label": "Last item",
        "content": "»",
      }
    }
    nextItem={
      Object {
        "aria-label": "Next item",
        "content": "⟩",
      }
    }
    onPageChange={[Function]}
    pageItem={Object {}}
    prevItem={
      Object {
        "aria-label": "Previous item",
        "content": "⟨",
      }
    }
    siblingRange={1}
    totalPages={0}
  />
</div>
`;
